### ISSUE TITLE
tests/codetoanalyze/java/racerd false positives/failures?

### ISSUE THREAD

**Issue by korDen** [facebook/infer#1146]
I tried running a few tests from infer/tests/codetoanalyze/java/racerd locally, and many seem to fail:

For example, https://github.com/facebook/infer/blob/master/infer/tests/codetoanalyze/java/racerd/Annotations.java:

warning: THREAD_SAFETY_VIOLATION
Unprotected write. Non-private method `int Annotations.FP_functionalAcrossBoxingLongOk()` writes to field `this.mBoxedLong2` outside of synchronization.
 Reporting because the current class is annotated `@ThreadSafe`, so we assume that this method can run in parallel with other non-private methods in the class (including itself).
```
  344.     public int FP_functionalAcrossBoxingLongOk() {
  345.       if (b) {
  346. >       mBoxedLong2 = returnLong();
  347.       }
  348.       return 2;
```

warning: THREAD_SAFETY_VIOLATION
  Unprotected write. Non-private method `boolean Annotations.FP_functionalAcrossUnboxingOk()` writes to field `this.mBool2` outside of synchronization.
 Reporting because the current class is annotated `@ThreadSafe`, so we assume that this method can run in parallel with other non-private methods in the class (including itself).
```
  314.     public boolean FP_functionalAcrossUnboxingOk() {
  315.       if (b) {
  316. >       mBool2 = returnBoxedBool();
  317.       }
  318.       return b;
```

warning: THREAD_SAFETY_VIOLATION
  Unprotected write. Non-private method `void ExtendsClassOnUiThread.fooOk()` writes to field `this.f` outside of synchronization.
 Reporting because a superclass `class codetoanalyze.java.checkers.AllMethodsOnUiThread` is annotated `@ThreadSafe`, so we assume that this method can run in parallel with other non-private methods in the class (including itself).
```
  436.     @Override
  437.     void fooOk() {
  438. >     f = 9;
  439.     }
  440.   
```

My understanding is that methods ending Ok() should pass not raise a warning and methods ending Bad() should raise a warning. Is the test wrong?

**Comment by ngorogiannis** [comment]
Indeed, that's the intuition behind the naming of the methods in the test.

I can't reproduce these failures.  Is this on master or a downloaded release?  Can you share how you run Infer? 

Can you run `make direct_java_racerd_test` from the root directory of the infer checkout and report the result?

**Comment by korDen** [comment]
I could not reproduce it with make direct_java_racerd_test either: 
[12:11:28][82539] Running test: direct java racerd test...
[      3s][82539] SUCCESS Running test: direct java racerd test

Turns out, the errors are whitelisted, which is why make direct_java_racerd_test is ignoring them: https://github.com/facebook/infer/blob/master/infer/tests/codetoanalyze/java/racerd/issues.exp

The way I tested it was making a BUCK file in the test directory and commented out the Bad() methods since all of them are meant to fail anyway. The BUCK file I used:

java_library(
  name = 'racerd',
  srcs = [
    'Annotations.java', # Annotations.java is referencing class Obj
    'Ownership.java', # which is defined in Ownership.java which references class SkippedClass
    'SkippedClass.java', # so we have to include all of these
  ],
  deps = [
    '//infer/annotations:annotations',
    '//dependencies/java/jsr-330:jsr-330',
    '//dependencies/java/jsr-305:jsr-305',
    '//dependencies/java/android/support/v4:android-support-v4',
  ],
)

I dug a little deeper and found that the methods were added and immediately whitelisted in this commit https://github.com/facebook/infer/commit/0e77e3623580475860a4bcad07613e3e8ed17db4 (D4501552)

So I think it's either a bug that they were whitelisted, mistake in naming and the methods should be called Bad() (unlikely), or perhaps missing functionality (e.g. they were meant to be passing but the functionality wasn't there to make them pass, and was never added).

**Comment by korDen** [comment]
I believe the Okay methods should pass. Here is the first failure in question:

```
Long mBoxedLong;

@Functional
native Long returnBoxedLong();

@Functional
native long returnLong();

public int functionalBoxedLongOk() {
  if (b) {
    mBoxedLong = returnBoxedLong(); // passes
  }
  return 22;
}

public int FP_functionalAcrossBoxingLongOk() {
  if (b) {
    mBoxedLong2 = returnLong(); // fails
  }
  return 2;
}
```

There is a race, but it's benign. The pattern is otherwise known as 'LazyInit', so let me add a quote:
https://github.com/google/error-prone/blob/master/annotations/src/main/java/com/google/errorprone/annotations/concurrent/LazyInit.java

"Use this annotation on any static or field that will be initialized lazily, where races yield no semantic difference in the code (as, for example, is the case with {@link String#hashCode})."

"Also note that LazyInit must not be used on 64-bit primitives (longs and doubles), because the Java Language Specification does not guarantee that writing to these is atomic. Furthermore, when used for non-primitives, the non-primitive must be either truly immutable or at least thread safe (in the Java memory model sense). And callers must accommodate the fact that different calls to something like the above getData() method may return different (though identically computed) objects, with different identityHashCode() values."

As you can see, `returnBoxedLong()` is marked `@Functional` so even if there is a race condition, as long as the write is atomic, the order doesn't matter (the race yields no semantic difference in the code). In the test above, returnBoxedLong() returns a Long reference, which can be written atomically and the test it's passing.

However, returnLong() is NOT passing despite being `@Functional`. The reason is likely that boxing doesn't automatically inherit `@Functional` attribute. Not sure if it's supposed to (maybe it was meant to, but was never implemented). The likely issue here is that instance may or may not be the same when boxing a Long, and `@Functional` is designed to return /the same/ value: https://github.com/facebook/infer/blob/master/infer/annotations/src/main/java/com/facebook/infer/annotation/Functional.java "Annotation for methods that always return the same value."

Relevant part from LazyInit Javadoc: "Furthermore, when used for non-primitives, the non-primitive must be either truly immutable or at least thread safe (in the Java memory model sense). And callers must accommodate the fact that different calls to something like the above getData() method may return different (though identically computed) objects, with different identityHashCode() values."

So in this case, `@Functional` is more strict than `@LazyInit`. Personally, I think that requiring `@Functional` to return the same Object is too prohibitive for real use, and essentially limits its use to returning primitive types or preallocated objects.

Either way, this test may be legitimately broken by design and should be called Bad().

The other one likely fails for a similar (yet reverse) reason, and I think should pass:

```
@Functional
native Boolean returnBoxedBool();

boolean mBool2;

public boolean FP_functionalAcrossUnboxingOk() {
    if (b) {
      mBool2 = returnBoxedBool(); // fails because unboxing doesn't retain @Functional attribute.
    }
    return b;
}

```
Here, returnBoxedBool() returns the same instance always, and unboxing it will return the same value always but `@Functional` attribute is lost in translation, triggering a warning.

Hope that helps.

**Comment by ngorogiannis** [comment]
Thanks for the detailed report!  FYI in tests where we have "FP" in the method name, that indicates we know this is a False Positive.  I'll look into this, thanks again.
